Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -255,36 +310,230 @@ public actor KafkaProducer {
-
-        return KafkaProducerMessageID(rawValue: self.messageIDCounter)
     }
 
     /// Check `topicHandles` for a handle matching the topic name and create a new handle if needed.
     /// - Parameter topic: The name of the topic that is addressed.
-    private func createTopicHandleIfNeeded(topic: String) throws -> OpaquePointer? {
-        if let handle = self.topicHandles[topic] {
-            return handle
-        } else {
-            let newHandle = try self.client.withKafkaHandlePointer { handle in
-                let rdTopicConf = try RDKafkaTopicConfig.createFrom(topicConfig: self.topicConfig)
-                return rd_kafka_topic_new(
-                    handle,
-                    topic,
-                    rdTopicConf
+    private func _createTopicHandleIfNeeded(topic: String) throws -> OpaquePointer? {
+        try self.stateMachine.withLockedValue { state in
+            let action = try state.createTopicHandleIfNeeded(topic: topic)
+            switch action {
+            case .handleExists(let handle):
+                return handle
+            case .createTopicHandle(let client, let topic):
+                let newHandle = try client.withKafkaHandlePointer { handle in
+                    let rdTopicConf = try RDKafkaTopicConfig.createFrom(topicConfig: self.topicConfig)
+                    return rd_kafka_topic_new(
+                        handle,
+                        topic,
+                        rdTopicConf
+                    )
+                    // rd_kafka_topic_new deallocates topic config object
+                }
+                if let newHandle {
+                    try state.addTopicHandle(topic: topic, handle: newHandle)
+                }
+                return newHandle
+            }
+        }
+    }
+}
+
+// MARK: - KafkaProducer + StateMachine
+
+extension KafkaProducer {
+    /// State machine representing the state of the ``KafkaProducer``.
+    struct StateMachine {
+        /// A logger.
+        let logger: Logger
+
+        /// The state of the ``StateMachine``.
+        enum State {
+            /// The state machine has been initialized with init() but is not yet Initialized
+            /// using `func initialize()` (required).
+            case uninitialized
+            /// The ``KafkaProducer`` has started and is ready to use.
+            ///
+            /// - Parameter messageIDCounter:Used to incrementally assign unique IDs to messages.
+            /// - Parameter client: Client used for handling the connection to the Kafka cluster.
+            /// - Parameter source: ``NIOAsyncSequenceProducer/Source`` used for yielding new elements.
+            /// - Parameter topicHandles: Dictionary containing all topic names with their respective `rd_kafka_topic_t` pointer.
+            case started(
+                client: KafkaClient,
+                messageIDCounter: UInt,
+                source: Producer.Source?,
+                topicHandles: [String: OpaquePointer]
```

Comment:
Could we abstract the handles into a class as well so we avoid the `OpaquePointer` here?